pkg/guestagent/iptables/iptables.go

```diff
@@ -0,0 +1,111 @@
+package iptables
+
+import (
+	"bytes"
+	"errors"
+	"net"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type Entry struct {
+	IP   net.IP
+	Port int
+}
+
+// This regex can detect a line in the iptables added by portmap to do the
+// forwarding. The following two are examples of lines (notice that one has the
+// destinateion IP and the other does not):
+//    -A CNI-DN-2e2f8d5b91929ef9fc152 -d 127.0.0.1/32 -p tcp -m tcp --dport 8081 -j DNAT --to-destination 10.4.0.7:80
+//    -A CNI-DN-04579c7bb67f4c3f6cca0 -p tcp -m tcp --dport 8082 -j DNAT --to-destination 10.4.0.10:80
+// The -A on the front is to amend the rule that was already created. portmap
+// ensures the rule is created before creating this line so it is always -A.
+// CNI-DN- is the prefix used for rule for an individual container.
+// -d is followed by the IP address. The regular expression looks for a valid
+// ipv4 IP address. We need to detect this IP.
+// --dport is the destination port. We need to detect this port
+// -j DNAT this tells us it's the line doing the port forwarding.
+var findPortRegex = regexp.MustCompile(`.*-A\W+CNI-DN-\w*\W+(?:-d ((?:\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}))?(?:/\d+\W+)?-p.*--dport (\d+) -j DNAT`)
```

Comment:
Some nit-picking:

* Remove the leading `.*` because it isn't anchored to the beginning of the line.
* Replace all `\W` with `\s` to match only whitespace. E.g. you don't want to match an inverted rule like `! -d 127.0.0.1/32`, but `\W` does match the `!`.
* Added a space after the `-p`.

```suggestion
var findPortRegex = regexp.MustCompile(`-A\s+CNI-DN-\w*\s+(?:-d ((?:\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}))?(?:/\d+\s+)?-p .*--dport (\d+) -j DNAT`)
```

I suspect all `\s+` sequences can be replaced by single spaces (at least it still passes your tests):

```suggestion
var findPortRegex = regexp.MustCompile(`-A CNI-DN-\w* (?:-d ((?:\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}))?(?:/\d+ )?-p .*--dport (\d+) -j DNAT`)
```

I'm also tempted to say that `/\d+ ` should be replaced by `/32 ` because I don't know what this would mean for any other mask value. But then I don't think portmap will specify any mask, so it may not matter.

Reply:
> Remove the leading .* because it isn't anchored to the beginning of the line.

oops, at one point it was anchored with `^` but I removed that and failed to remove the `.*` along with it. Fixed.

> Replace all \W with \s to match only whitespace. E.g. you don't want to match an inverted rule like ! -d 127.0.0.1/32, but \W does match the !.

Great catch on the ! case. Fixed.

> Added a space after the -p.

Updated

> I suspect all \s+ sequences can be replaced by single spaces (at least it still passes your tests):

I used the `\s+` to be on the safe side. I was attempting to be loose in case something added more space in there. It should work with a single space. I'm leaving it that way for now to be loose with the parsing. If I really do need to tighten it please let me know.

> I'm also tempted to say that /\d+ should be replaced by /32 because I don't know what this would mean for any other mask value. But then I don't think portmap will specify any mask, so it may not matter.

I discovered the `/32` in my testing. I did not track down exactly what was adding it. I've updated it to only look for `/32` as that's the single IP and we should not have a range.